compiler/rustc_resolve/src/ident.rs

```diff
@@ -1369,6 +1375,22 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         let mut allow_super = true;
         let mut second_binding = None;
 
+        // Set aside the previously generated privacy errors so that we can extend any privacy
+        // errors being generated here.
+        // CAUTION! If you add a new return path to this function, you need to ensure that you
+        // recover these into it again, otherwise rustc will start allowing use of items that
+        // should remain unreachable.
+        let mut privacy_errors = std::mem::take(&mut self.privacy_errors);
```

Comment:
I think something like
```rust
let num_privacy_errors = self.privacy_errors.len();
// ...
for error in &mut self.privacy_errors[num_privacy_errors..] {
    error.outermost_res = last_res;
}
```
would help to avoid all this moving in/out both here and in `imports.rs`.

Reply:
I considered that, but I was concerned when I saw some logic that modified `privacy_errors`. Doing it that way provides no recourse to any change to the ordering of `privacy_errors` or removal of elements that invalidate the range. If this not happening is something that we feel comfortable relying on, maybe only with warnings on the field definition itself, then doing it that way is valid.

Reply:
I think its fine, `privacy_erros` are supposed to be append-only.

The only `privacy_errors.truncate(..)` is a recent addition (https://github.com/rust-lang/rust/pull/109788) for another error deduplication, maybe it's better to replace it with some span-based deduplication too to keep the append-only property (not in this PR).
It's "local" in the sense that we are popping what we just pushed, so it's not visible from the outside of `resolve_qpath`.